### PR TITLE
RDM-3460 Search and Workbasket changes to support complex fields

### PR DESCRIPTION
**Pull request by rafalkalita**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-3460

### Change description ###
Complex element *only* as list screen input.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java [hunk 1]
```diff
@@ -153,12 +152,17 @@ public void setSearchAliasFields(List<SearchAliasField> searchAliasFields) {
     }
 
     @JsonIgnore
-    public Optional<CaseField> getCaseField(String caseFieldId) {
-        return caseFields.stream().filter(caseField -> caseField.getId().equalsIgnoreCase(caseFieldId)).findFirst();
+    public CaseField getCaseField(String caseFieldId) {
+        return caseFields.stream().filter(caseField -> caseField.getId().equalsIgnoreCase(caseFieldId)).findFirst()
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java]: can I know what's the rationale for this change? I think the previous solution is cleaner. I don't see why we need to throw an Exception here

File: src/main/java/uk/gov/hmcts/ccd/domain/model/search/Field.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 public class Field {
     private String id;
+    private String elementPath;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/search/Field.java]: Do we need to add elementPath to Field? it's a quite specific property needed only for this particular use case. We already have added it to WorkbasketInputField and SearchInputField can't we use that property?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultFindSearchInputOperation.java [hunk 3]
```diff
@@ -51,19 +51,15 @@ private SearchInput toSearchInput(final SearchInputField in, final CaseType case
         final SearchInput result = new SearchInput();
         result.setLabel(in.getLabel());
         result.setOrder(in.getDisplayOrder());
+        CaseField caseField = caseType.getCaseField(in.getCaseFieldId());
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultFindSearchInputOperation.java]: you can use the original `getCaseField` method that returns Optional here and throw exception at this level

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java [hunk 4]
```diff
@@ -140,6 +144,46 @@ public void propagateACLsToNestedFields() {
...
+        if (StringUtils.isBlank(path)) {
+            return this.getFieldType();
+        } else {
+            CaseField caseFieldByPath = this.findNestedElementByPath(path);
+            return caseFieldByPath.getFieldType();
+        }
+    }
+
+    @JsonIgnore
+    public CaseField findNestedElementByPath(String path) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: can this be private? In that case we could remove the blank field validation because we already did it in `getFieldTypeByPath`

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java [hunk 5]
```diff
@@ -140,6 +144,46 @@ public void propagateACLsToNestedFields() {
         propagateACLsToNestedFields(this, this.accessControlLists);
     }
 
+    @JsonIgnore
+    public FieldType getFieldTypeByPath(String path) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: I might be wrong but such a method would be better placed in the FieldType class rather than CaseField. You could return an Optional<FieldType>. And the following check:
`if (StringUtils.isBlank(path)) {
            return this.getFieldType();
 }`
could be removed from here and placed in the `DefaultFindSearchInputOperation`

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java [hunk 6]
```diff
@@ -140,6 +144,46 @@ public void propagateACLsToNestedFields() {
...
+        }
+    }
+
+    @JsonIgnore
+    public CaseField findNestedElementByPath(String path) {
+        if (StringUtils.isBlank(path)) {
+            throw new RuntimeException(format("Invalid blank element path for field %s.", this.id));
+        }
+        if (this.getFieldType().getChildren().isEmpty()) {
+            throw new RuntimeException(format("CaseField %s has no nested elements.", this.id));
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: I would expect the endpoint to return a 400 Bad Request exception when this.getFieldType().getChildren().isEmpty(). Can we double check that a RuntimeException thrown here gets mapped by the endpoint to a 400 status code and not a 50X type of error?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java [hunk 7]
```diff
@@ -140,6 +144,46 @@ public void propagateACLsToNestedFields() {
...
+
+        return reduce(this.getFieldType().getChildren(), pathElements);
+    }
+
+    @JsonIgnore
+    private CaseField reduce(List<CaseField> caseFields, List<String> pathElements) {
+        String head = pathElements.get(0);
+        if (pathElements.size() == 1) {
+            return caseFields.stream().filter(e -> e.getId().equals(head)).findFirst()
+                .orElseThrow(() -> new RuntimeException(format("Nested element not found for %s", head)));
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: can we return a more descriptive error message and also the `path` as part of the message?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java [hunk 8]
```diff
@@ -140,6 +144,46 @@ public void propagateACLsToNestedFields() {
...
+        } else {
+            CaseField caseField = caseFields.stream().filter(e -> e.getId().equals(head)).findFirst()
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: this block of code is duplicated with the above. Can we remove this duplication?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java (hunk 1 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java]: Changed back

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java (hunk 8 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: Done

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java (hunk 7 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: head is a part of the path. Changed the name to reflect it better.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java (hunk 6 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: Returning BadRequestException that is our exception mapped to 400

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java (hunk 5 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: Moved to CaseType. Now it is named getCaseFieldByPath(String caseFieldId, String path)

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java (hunk 4 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseField.java]: Now used by CaseType getCaseFieldByPath(...) method

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultFindSearchInputOperation.java (hunk 3 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultFindSearchInputOperation.java]: Changed as in above

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java [hunk 9]
```diff
@@ -153,12 +154,34 @@ public void setSearchAliasFields(List<SearchAliasField> searchAliasFields) {
...
+     * Gets a caseField by caseField id and specified path.
+     *
+     * @param caseFieldId CaseField id
+     * @param path Path to a nested CaseField
+     * @return Optional of a CaseField
+     */
     @JsonIgnore
-    public boolean isCaseFieldACollection(String caseFieldId) {
-        return getCaseField(caseFieldId).map(CaseField::isCollectionFieldType).orElse(false);
+    public Optional<CaseField> getCaseFieldByPath(String caseFieldId, String path) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java]: I think you can simplify the method to:
```
Optional<CaseField> caseFieldOptional = getCaseField(caseFieldId);

        if (StringUtils.isBlank(path)) {
            return caseFieldOptional;
        }
        return caseFieldOptional.map(f-> f.findNestedElementByPath(path));
```

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseType.java]: I'm just realising now that the path is relative to a field id, it's not an absolute path. In this case you were right this method belongs to the CaseField not the CaseType

**Comment by tzarsmango** [comment]
This PR has lots of conflicting changes with [this PR](https://github.com/hmcts/ccd-data-store-api/pull/453) Probably it would be better to review after merging the other one and refreshing this one.
